module5/lessons/Week3/Authorization.md

```diff
@@ -0,0 +1,212 @@
+
+#### Add Admin Role to Database
+
+<section class='call-to-action' markdown='1'>
+
+**Practice your SQL!**  
+
+* Open pgAdmin and connect to the IdentityMessageBoard Database
+* Insert a record into `AspNetRoles`
+    * name should be 'Admin'
+* Add the admin role to a user
+    * Hint: keep a close eye on your ERD! Make sure you are adding records to the correct table.  If you are unsure, ask an instructor!
+
+</section>
+
+<section class='instructor-notes' markdown='1'>
+
+SQL Query for assigning Admin to a user in the database (by id):
+
+```SQL
+INSERT INTO "AspNetRoles" (id, name)
+VALUES (1, 'Admin');
+
+INSERT INTO "AspNetUserRoles" (user_id, role_id)
+VALUES ('<a user id>', 1);
+```
+
+</section>
+
+When you have configured an admin role and assigned it to a user, that user should be able to see the All Messages View!
+
+<section class='note' markdown='1'>
+
+The `[Authorize]` decorator can be used on an entire controller, or on a specific action.  If you use it on an entire controller, you can exempt actions from authorization with the `[AllowAnonymous]` decorator!
+
+</section>
+
+## Claims Based Authorization
+
+Compared to Roles based authorization, claims are a bit trickier to manage. By definition, claims are things that describe a user (like age), that can change over time.  
+
+<section class='call-to-action' markdown='1'>
+
+With a partner, take a look at the ERD for our current program.  Answer the following:
+* What information about claims is stored?
+* How are claims related to users?
+* What does this mean for how developers must manage claims in ASP.NET Core Identity?
+
+</section>
+
+Out of the box, claims are stored as key/value pairs associated with specific users.  This could be ok for things like 'birthdate' where the key value pair could be `"birthdate" : "01/13/1999"`.  In this example, the age of a user will change, but the underlying data will not.
+
+But consider a more complex example.  What if we wanted to give users a different experience based on the number of times they interacted with our application.  Maybe we want them to have sparkly font if they have visited the site more than 100 times.  In that case, they key/value pair becomes more cumbersome - we would have to update that database record every time a user visits our page.
+
+The more we are manipulating records, the more likely we are to have developer-errors impact our clients.
+
+So, while there is funtionality included in the Identity Framework to handle Claim based authorization, it might not be the best.
+
+<section class='call-to-action' markdown='1'>
+
+In small groups, design a solution for the following scenario.  You do not need to implement the solution just yet, but come up with some high-level structures for how you might achieve this goal.
```

Comment:
I'm not following what you're going with for another another approach besides key value pairs. Updating the value for the particular key for every visit feels to me like the natural way to do this? And then that key and value it looks like are written to fields in the claimUser table, which also seems to make sense. 

Can you add more details on what you're thinking students will come up with and implement for another approach?

Reply:
Oh this is so interesting.  To me, storing this kind of 'changeable' data as a key/value feels very un-natural - like duplicating the storage of information.  For me, the natural feeling way is to use the data we are already storing on a user (.Messages) to check how many messages each user has.  When a message is created, we are already storing the information we need to make this authorization decision.  If we **also** had a key/value pair, we make the creation of a message more complicated (we have to update the claim pair) without gaining much benefit (we already have the information we need to make this decision).

There is a "simple" implementation that students could use where we check a user's message count in the controller action (or view), and generate the easter egg from there.  Or a more complex implementation (that I don't expect students to do) where we create a [custom authorization policy based on a resource](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/resourcebased?view=aspnetcore-7.0)

I'll add some instructor notes around the different implementations we might see - I don't think there is anything wrong with implementing a true claim here; I just wouldn't make that choice 😆